lib/smb2/tree.rb

```diff
@@ -28,23 +34,68 @@ def initialize(client:, tree_connect_response:)
   # The protocol is persnickety about format. The `filename` must not begin
   # with a backslash or it will return a STATUS_INVALID_PARAMETER.
   #
+  # If the optional code block is given, it will be passed the opened file as
+  # an argument and the {File} object will automatically be {File#close
+  # closed} when the block terminates. The value of the block will be
+  # returned.
+  #
   # @param filename [String,#encode] this will be encoded in utf-16le
+  # @param mode [String] See stdlib ::File#mode
+  # @yield [Smb2::File]
   # @return [Smb2::File]
-  def create(filename)
-    packet = Smb2::Packet::CreateRequest.new do |request|
-      request.filename = filename.encode("utf-16le")
-      # @todo document all these flags. value copied from smbclient traffic
-      request.desired_access = 0x0012_0089
-      request.impersonation = 2
-      request.share_access = 3  # SHARE_WRITE | SHARE_READ
-      request.disposition = 1   # Open (if file exists open it, else fail)
-      request.create_options = 0x40 # non-directory
-    end
+  def create(filename, mode = "r+")
+    desired_access = desired_access_from_mode(mode)
+    create_options = Smb2::Packet::CREATE_OPTIONS[:FILE_NON_DIRECTORY_FILE]
+
+    packet = Smb2::Packet::CreateRequest.new(
+      filename: filename.encode("utf-16le"),
+      desired_access: desired_access,
+      impersonation: 2,
+      share_access: 3,  # SHARE_WRITE | SHARE_READ
+      disposition: disposition_from_file_mode(mode),
+      create_options: create_options
+    )
 
     response = send_recv(packet)
 
     create_response = Smb2::Packet::CreateResponse.new(response)
-    Smb2::File.new(tree: self, create_response: create_response)
+    file = Smb2::File.new(filename: filename, tree: self, create_response: create_response)
+    if mode.start_with?("a")
+      file.seek(create_response.end_of_file)
+    end
+
+    if block_given?
+      value = yield file
+      file.close
+      value
+    else
+      file
+    end
+  end
+
+  def delete(filename)
+    packet = Smb2::Packet::CreateRequest.new(
+      filename: filename.encode("utf-16le"),
+      desired_access: Smb2::Packet::FILE_ACCESS_MASK[:FILE_DELETE],
+      impersonation: 2,
+      share_access: 7, # SHARE_DELETE | SHARE_WRITE | SHARE_READ
```

Comment:
Please define the constants in the comment and `|` together instead of hard-coding `7` as a magic number.